### PR TITLE
GH-38709: [C++] Protect against PREALLOCATE preprocessor defined on macOS

### LINKED ISSUE #38709: [C++] Conflict between `arrow::compute::MemAllocation::PREALLOCATE` and system header `vnode.h` preprocessor define
### Describe the bug, including details regarding any error messages, version, and platform.

Hi, 

On MacOSX the system header [`vnode.h`](https://opensource.apple.com/source/xnu/xnu-6153.41.3/bsd/sys/vnode.h.auto.html) defines the symbol `PREALLOCATE` as a preprocessor define.   This causes a compilation error with `#include <arrow/compute/kernel.h>` where this name (`PREALLOCATE`) is used as an identifier , if `sys/vnode.h` and `arrow/compute/kernel.h` are both (possibly indirectly) included into the same compilation unit. 

Perhaps the `PREALLOCATE` symbol should be `#undef` in `arrow/compute/kernel.h` in case it is defined as preprocessor macro - e.g., 

```
    #if defined(__APPLE__) and defined(PREALLOCATE)
    # undef PREALLOCATE 
    #endif  
```
Yours,
_Christian_

### Component(s)

C++

### PR DESCRIPTION
**Pull request by cholmcc**
### Rationale for this change

The macOS header `sys/vnode.h` defines `PREALLOCATE` as a preprocessor macro, which causes a conflict in `arrow/compute/kernel.h` where the same name is defined as an identifier. 

Other BSDs does not seem to define this macro. 

### What changes are included in this PR?

`#undef PREALLOCATE` on macOS and if defined. 

### Are these changes tested?

Somewhat, in a different context. 

### Are there any user-facing changes?

If some code specific to macOS actually uses the `PREALLOCATE` macro, then that code may have problems.  However, it is unlikely that any user code would use this macro as it seems to be intended for internal use in the BSD kernel of macOS. 

* Closes: #38709

### CODE REVIEW

File: cpp/src/arrow/compute/kernel.h [hunk 1]
```diff
@@ -37,6 +37,11 @@
 #include "arrow/type.h"
 #include "arrow/util/macros.h"
 #include "arrow/util/visibility.h"
+// MacOSX defines PREALLOCATE as a preprocessor define in the header sys/vnode.h
+// No other BSD seems to do so.  The name is used as an identifier in MemAllocation enum
```
Comment (kou) [inline cpp/src/arrow/compute/kernel.h:41]:
```suggestion

// macOS defines PREALLOCATE as a preprocessor macro in the header sys/vnode.h.
// No other BSD seems to do so. The name is used as an identifier in MemAllocation enum.
```

**Review by kou — APPROVED** [review]
+1

File: cpp/src/arrow/compute/kernel.h [hunk 2]
```diff
@@ -38,6 +38,12 @@
 #include "arrow/util/macros.h"
 #include "arrow/util/visibility.h"
 
+// macOS defines PREALLOCATE as a preprocessor macro in the header sys/vnode.h.
+// No other BSD seems to do so. The name is used as an identifier in MemAllocation enum.
+#if defined(__APPLE__) && defined(PREALLOCATE)
+# undef PREALLOCATE 
+#endif
```
Comment (kou) [inline cpp/src/arrow/compute/kernel.h:45]:
```suggestion
#undef PREALLOCATE
#endif
```

**Comment by pitrou** [comment]
Sidenote: this is why we should discourage ENUM_VALUE in favor of `kEnumValue` in new code :-)

**Comment by kou** [comment]
Oh, I forgot to merge this.
I'll merge this.

**Pull request merged**